impl/src/main/java/org/jboss/weld/manager/BeanManagerImpl.java

```diff
@@ -1559,4 +1569,160 @@ public void validationFailed(Exception failure, Environment environment) {
+                }
+            }
+            Set<Bean<?>> removable = new HashSet<>();
+            Set<Bean<?>> unusedProducers = new HashSet<>();
+            WeldUnusedMetadataExtension metadataExtension = getUnusedMetadataExtension();
+
+            for (Bean<?> bean : enabledBeans) {
+                if (bean instanceof ForwardingBean) {
+                    // Unwrap the forwarding implementation
+                    ForwardingBean<?> forwarding = (ForwardingBean<?>) bean;
+                    bean = forwarding.delegate();
+                }
+                // Built-in bean, extension, interceptor, decorator, session bean
+                if (bean instanceof AbstractBuiltInBean || bean instanceof ExtensionBean || bean instanceof Interceptor || bean instanceof Decorator
+                        || bean instanceof SessionBean) {
+                    continue;
+                }
+                // Has a name
+                if (bean.getName() != null) {
+                    continue;
+                }
+                // The type is excluded
+                if (excludeType != null && excludeType.matcher(bean.getBeanClass().getName()).matches()) {
+                    continue;
+                }
+                // Declares an observer
+                if (beanToDeclaredObservers.containsKey(bean)) {
+                    continue;
+                }
+                // Declares a used producer - see also second pass
+                if (beanToDeclaredProducers.containsKey(bean)) {
+                    continue;
+                }
+                // Is resolved for an injection point
+                if (validator.isResolved(bean) || (metadataExtension != null
+                        && (metadataExtension.isInjectedByEEComponent(bean, this) || metadataExtension.isInstanceResolvedBean(bean, this)))) {
+                    continue;
+                }
+                // The type is annotated with an exclude annotation
+                if (excludeAnnotation != null && hasExcludeAnnotation(bean, excludeAnnotation)) {
+                    continue;
+                }
+                // This bean is very likely an unused producer
+                if (bean instanceof AbstractProducerBean) {
+                    unusedProducers.add(bean);
+                }
+                BootstrapLogger.LOG.dropUnusedBeanMetadata(bean);
+                removable.add(bean);
+            }
+
+            // Second pass to find beans which themself are unused and declare only unused producers
+            if (!unusedProducers.isEmpty()) {
+                for (Bean<?> bean : enabledBeans) {
+                    if (bean instanceof ForwardingBean) {
+                        // Unwrap the forwarding implementation
+                        ForwardingBean<?> forwarding = (ForwardingBean<?>) bean;
+                        bean = forwarding.delegate();
+                    }
+                    if (!beanToDeclaredProducers.containsKey(bean)) {
```

Comment:
Couldn't we just loop straight over `beanToDeclaredProducers.keySet()`?

Reply:
+10 ;-)